src/dataset.jl

```diff
@@ -172,13 +172,111 @@ end
     dataset::AbstractDataset,
     filename::AbstractString;
     kwargs...,
 )::Nothing
-    destroy(unsafe_copy(dataset, filename = filename; kwargs...))
+    if nraster(dataset) > 0 && nlayer(dataset) > 0
+        drivername = :driver ∈ keys(kwargs) ? shortname(kwargs[:driver]) : shortname(getdriver(dataset))
+        error("Writing datasets with raster and vector data is not supported when using driver $drivername. 
+        Please file an issue at https://github.com/yeesian/ArchGDAL.jl/issues 
+        including following dataset information: \n\n$dataset")
+    elseif nraster(dataset) > 0
+        destroy(unsafe_copy(dataset, filename = filename; kwargs...))
+    elseif nlayer(dataset) > 0
+        writelayers(dataset, filename; kwargs...)
+    end
+    return nothing
+end
+
+"""
+    writelayers(ds, filename; driver, options=[""], layer_options=[""], chunksize=20000)
+
+Writes the vector dataset to the designated filename. The options are passed to the newly created dataset and
+have to be given as a list of strings in KEY=VALUE format. The chunksize controls the number of features written 
+in each database transaction, e.g. for SQLite. This function can also be used to copy datasets on disk.
+
+Currently working drivers: FlatGeobuf, GeoJSON, GeoJSONSeq, GML, GPKG, JML, KML, MapML, ESRI Shapefile, SQLite
+
+### Parameters
+*`dataset`  The source dataset
+*`filename` The file name to write to
+
+### Keyword arguments
+`driver`           The driver to use, you have to manually select the right driver for the file extension you wish
+`options`   A vector of strings containing KEY=VALUE pairs for driver-specific creation options
+`layer_options`    Driver specific options for layer creation
+`chunksize`        Number of features to write in one database transaction. Neglected when `use_gdal_copy` is true.
+`use_gdal_copy`    Set this to true if you encounter errors when writing. Set to false for higher speed.
+
+### Returns
+nothing
+"""
+function writelayers(dataset::AbstractDataset,
+                     filename;
+                     driver=getdriver(dataset),
+                     options::Vector{String}=[""], 
+                     layer_options::Vector{String}=[""],
+                     chunksize=20000,
+                     use_gdal_copy=false)
+    create(filename; driver=driver, options=options) do target
+        for layeridx in 0:nlayer(dataset)-1
+            sourcelayer = getlayer(dataset, layeridx)
+            sourcelayerdef = layerdefn(sourcelayer)
+            
+            if shortname(driver) == "GPKG"
+                if !any(Base.contains.(layer_options, "GEOMETRY_NAME"))
+                    if ngeom(sourcelayer) > 0  # for GPKG there can be only one geometry column / layer
+                        geometry_column_name = getname(getgeomdefn(sourcelayerdef, 0))
+                        push!(layer_options, "GEOMETRY_NAME=$geometry_column_name")
+                    end
+                end
+            end
```

Comment:
This looks much better to me compared to the previous approach -- I'm trying to vet with what's being done in https://github.com/Toblerity/Fiona and why they didn't have similar issues for GPKG for my own general understanding. Can you comment on what's happening with layer options here?

Reply:
> Can you comment on what's happening with layer options here?

Okay I think I understand now, sorry it took me this long. My remaining question for this thread (for my understanding) is whether we would have the right behavior for GPKG if `use_gdal_copy=true`? (My guess is "yes", but I'd like to be sure here.)

Reply:
I looked up [the corresponding implementation in GDAL](https://github.com/OSGeo/gdal/blame/7a4c4903428d289c5a63f51d5175f9564fb5d5d7/gcore/gdaldataset.cpp#L5106-L5447): we should 
1. explain that we haven't [handled the coordination transformations](https://github.com/OSGeo/gdal/blame/7a4c4903428d289c5a63f51d5175f9564fb5d5d7/gcore/gdaldataset.cpp#L5209-L5222)
2. default to `use_gdal_copy=true` (which is the correct; albeit slower implementation), unless we decide to port the GDAL implementation over and clean up the edge cases and make it correct here. We should still maintain the warning that users might encounter errors if they set `use_gdal_copy=false` (but it will be faster).

Reply:
> I think the basic question is, why it is slower than my implementation. I guess this is due to the low count of features per database transaction (128 vs 20000). So it might be worthwhile to file a feature request with gdal to introduce this as a parameter to `CopyLayer` or so.

Yeah that was what I found too: https://github.com/Toblerity/Fiona/pull/491#discussion_r148936648 and I agree with your assessment about filing the feature request for it to be a parameter rather than hardcoded constant.

> In the meantime, we can cook our own soup and I will:
> 
> 1. explain the missing possibility of coordinate transformations when use_gdal_copy is false in the docstring(before merge)
> 2. set the default of use_gdal_copy to true (before merge)
> 3. work on the julia implementation to get rid of the need for the workaround altogether (after merge?)
> 4. implement the coord transforms in julia (after merge?)

Yeah, that works for me, thank you!